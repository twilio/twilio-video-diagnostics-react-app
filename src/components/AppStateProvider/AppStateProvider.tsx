import React, { createContext, useContext, useReducer, useEffect } from 'react';
import produce from 'immer';
import { VideoInputTest } from '@twilio/rtc-diagnostics';

export enum ActivePane {
  GetStarted,
  DeviceCheck,
  DeviceError,
  CameraTest,
  Connectivity,
  Quality,
  Results,
}

interface stateType {
  activePane: ActivePane;
  videoGranted: boolean;
  audioGranted: boolean;
  deviceError: null | Error;
  videoInputTestReport: null | VideoInputTest.Report;
}

type ACTIONTYPE =
  | { type: 'set-active-pane'; newActivePane: ActivePane }
  | { type: 'next-pane' }
  | { type: 'previous-pane' }
  | { type: 'set-devices'; devices: MediaDeviceInfo[] }
  | { type: 'set-device-error'; error: Error }
  | { type: 'set-video-test-report'; report: VideoInputTest.Report };

type AppStateContextType = {
  state: stateType;
  dispatch: React.Dispatch<ACTIONTYPE>;
};

export const initialState = {
  activePane: ActivePane.GetStarted,
  videoGranted: false,
  audioGranted: false,
  deviceError: null,
  videoInputTestReport: null,
};

export const AppStateContext = createContext<AppStateContextType>(null!);

export function useAppStateContext() {
  const context = useContext(AppStateContext);
  if (!context) {
    throw new Error('useAppStateContext must be used within an AppStateProvider');
  }
  return context;
}

/**
 * This reducer function helps with the logic used to decide the activePane,
 * as well as which panes are to be shown to the user, and when.
 */

export const appStateReducer = produce((draft: stateType, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'set-active-pane':
      draft.activePane = action.newActivePane;
      break;

    case 'next-pane':
      switch (draft.activePane) {
        case ActivePane.GetStarted:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.CameraTest;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        case ActivePane.DeviceCheck:
          if (draft.deviceError) {
            draft.activePane = ActivePane.DeviceError;
          } else {
            draft.activePane = ActivePane.CameraTest;
          }
          break;
        default:
          draft.activePane++;
          break;
      }
      break;

    case 'previous-pane':
      switch (draft.activePane) {
        case ActivePane.CameraTest:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.GetStarted;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        default:
          draft.activePane--;
          break;
      }
      break;

    case 'set-devices':
      draft.audioGranted = action.devices.filter((d) => d.kind === 'audioinput').every((d) => d.label);
      draft.videoGranted = action.devices.filter((d) => d.kind === 'videoinput').every((d) => d.label);
      break;

    case 'set-device-error':
      draft.deviceError = action.error;
      draft.activePane = ActivePane.DeviceError;
      break;

    case 'set-video-test-report':
      draft.videoInputTestReport = action.report;
  }
});

export const AppStateProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(appStateReducer, initialState);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then((devices) => dispatch({ type: 'set-devices', devices }));
  }, []);

  return <AppStateContext.Provider value={{ state, dispatch }}>{children}</AppStateContext.Provider>;
};
