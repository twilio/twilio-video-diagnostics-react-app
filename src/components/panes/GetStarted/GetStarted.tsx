import { useAppStateContext } from '../../AppStateProvider/AppStateProvider';
import { Button, Container, Grid, Typography } from '@material-ui/core';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Hello from './Hello.png';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    gridContainer: {
      justifyContent: 'space-between',
    },
    /* 
    The size of the image is explicitly stated here so that this content can properly be centered vertically
    before the image is loaded.
    */
    illustration: {
      width: '284px',
      height: '284px',
      [theme.breakpoints.down('md')]: {
        width: '200px',
        height: '200px',
      },
    },
  })
);

export function GetStarted() {
  const { nextPane } = useAppStateContext();
  const classes = useStyles();

  return (
    <Container>
      <Grid container alignItems="center" className={classes.gridContainer}>
        <Grid item lg={5}>
          <Typography variant="h1" gutterBottom>
            Let's get started.
          </Typography>

          <Typography variant="body1" gutterBottom>
            We'll help you solve any video troubles you're experiencing but first, let's check your setup.
          </Typography>

          <Button variant="contained" color="primary" onClick={nextPane} style={{ marginBottom: '1em' }}>
            Get started
          </Button>
        </Grid>

        <Grid item lg={5}>
          <img src={Hello} alt="Hello" className={classes.illustration} />
          <Typography variant="body1" color="textSecondary">
            <strong>Not sure about something?</strong> Skip that section for now, and your support administrator can
            help later.
          </Typography>
        </Grid>
      </Grid>
    </Container>
  );
}
