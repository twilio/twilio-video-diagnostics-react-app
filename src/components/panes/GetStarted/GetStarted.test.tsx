import { shallow } from 'enzyme';
import { Button } from '@material-ui/core';
import { GetStarted } from './GetStarted';
import { useAppStateContext } from '../../AppStateProvider/AppStateProvider';

jest.mock('../../AppStateProvider/AppStateProvider');

const mockUseAppStateContext = useAppStateContext as jest.Mock<any>;
mockUseAppStateContext.mockImplementation(() => ({
  state: { appIsExpired: true },
}));

describe('the GetStarted component', () => {
  it('should render correctly when the app has expired', () => {
    const wrapper = shallow(<GetStarted />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should disable the Get Started button when the app has expired', () => {
    const wrapper = shallow(<GetStarted />);
    const getStartedButton = wrapper.find(Button);

    expect(getStartedButton.prop('disabled')).toBe(true);
  });

  it('should render correctly when the app is not expired', () => {
    mockUseAppStateContext.mockImplementationOnce(() => ({
      state: { appIsExpired: false },
    }));

    const wrapper = shallow(<GetStarted />);

    expect(wrapper).toMatchSnapshot();
  });
});
