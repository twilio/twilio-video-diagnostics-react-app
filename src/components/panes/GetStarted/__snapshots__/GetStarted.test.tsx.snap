// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the GetStarted component should render correctly when the app has expired 1`] = `
<WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="makeStyles-gridContainer-1"
    container={true}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h1"
      >
        <ErrorIcon />
        App has expired.
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        Please re-deploy the application and try again.
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={true}
        style={
          Object {
            "marginBottom": "1em",
          }
        }
        variant="contained"
      >
        Get started
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={5}
    >
      <img
        alt="Hello"
        className="makeStyles-illustration-2"
        src="test-file-stub"
      />
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        gutterBottom={true}
        variant="body1"
      >
        <strong>
          Not sure about something?
        </strong>
         Skip that section for now, and your support administrator can help later.
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;

exports[`the GetStarted component should render correctly when the app is not expired 1`] = `
<WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="makeStyles-gridContainer-1"
    container={true}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h1"
      >
        Let's get started.
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        We'll help you solve any video troubles you're experiencing but first, let's check your setup.
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={false}
        style={
          Object {
            "marginBottom": "1em",
          }
        }
        variant="contained"
      >
        Get started
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={5}
    >
      <img
        alt="Hello"
        className="makeStyles-illustration-2"
        src="test-file-stub"
      />
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        gutterBottom={true}
        variant="body1"
      >
        <strong>
          Not sure about something?
        </strong>
         Skip that section for now, and your support administrator can help later.
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;
